src/mlpack/tests/augmented_rnns_tasks_test.cpp

```diff
@@ -0,0 +1,291 @@
+  {
+    arma::mat input = predictors.at(0);
+    arma::mat output = labels.at(0);
+    size_t zeroCnt = 0, oneCnt = 0;
+    for (size_t i = 1; i < input.n_rows; i += 2)
+    {
+      size_t& addVar = (input.at(i, 0) == 0) ? zeroCnt : oneCnt;
+      ++addVar;
+    }
+    assert(oneCnt % zeroCnt == 0);
+    nRepeats = oneCnt / zeroCnt;
+  }
+  void Predict(arma::mat& predictors,
+               arma::mat& labels)
+  {
+    size_t seqLen = (predictors.n_rows / 2) / (nRepeats + 1);
+    size_t outputLen = nRepeats * seqLen;
+    assert(2 * (seqLen + outputLen) == predictors.n_rows);
+    labels.zeros(predictors.n_rows / 2, 1);
+    for (size_t i = 0; i < outputLen; ++i)
+    {
+      labels.at(seqLen+i) = predictors.at(2 * (i % seqLen));
+    }
+  }
+  void Predict(arma::field<arma::mat>& predictors,
+               arma::field<arma::mat>& labels)
+  {
+    size_t sz = predictors.n_elem;
+    labels = arma::field<arma::mat>(sz);
+    for (size_t i = 0; i < sz; ++i)
+    {
+      Predict(predictors.at(i), labels.at(i));
+    }
+  }
+
+ private:
+  size_t nRepeats;
+};
+
+// The dummy model that simply sorts the sequence.
+class HardCodedSortModel {
+ public:
+  HardCodedSortModel(size_t bitLen) : bitLen(bitLen) {}
+  void Train(arma::field<arma::mat>& predictors,
+             arma::field<arma::mat>& labels)
+  {
+    assert(predictors.n_elem == labels.n_elem);
+  }
+  void Predict(arma::mat& predictors,
+               arma::mat& labels)
+  {
+    predictors = predictors.t();
+    predictors.reshape(bitLen, predictors.n_elem / bitLen);
+    size_t len = predictors.n_cols;
+    labels.zeros(bitLen, len);
+    vector<pair<int, int>> vals(len);
+    for (size_t j = 0; j < len; ++j)
+    {
+      int val = 0;
+      for (size_t k = 0; k < bitLen; ++k)
+      {
+        val <<= 1;
+        val += predictors.at(k, j);
+      }
+      vals[j] = make_pair(val, j);
+    }
+    sort(vals.begin(), vals.end());
+    for (size_t j = 0; j < len; ++j)
+    {
+      labels.col(j) = predictors.col(vals[j].second);
+    }
+    labels.reshape(predictors.n_elem, 1);
+  }
+  void Predict(arma::field<arma::mat>& predictors,
+               arma::field<arma::mat>& labels)
+  {
+    auto sz = predictors.n_elem;
+    labels = arma::field<arma::mat>(sz);
+    for (size_t i = 0; i < sz; ++i)
+    {
+      Predict(predictors.at(i), labels.at(i));
+    }
+  }
+
+ private:
+  size_t bitLen;
+};
+
+// The dummy model that simply add two binary numbers.
+class HardCodedAddModel {
+ public:
+  HardCodedAddModel() {}
+  void Train(arma::field<arma::mat>& predictors,
+             arma::field<arma::mat>& labels)
+  {
+    return;
+  }
+  void Predict(arma::mat& predictors,
+               arma::mat& labels)
+  {
+    assert(predictors.n_elem % 3 == 0);
+    predictors = predictors.t();
+    predictors.reshape(3, predictors.n_elem / 3);
+    assert(predictors.n_rows == 3);
+    int num_A = 0, num_B = 0;
+    bool num = false; // true iff we have already seen the separating symbol
```

Comment:
No, I mean "iff" in the meaning "if an only if" - just like it is normally done in math texts